diff --git a/ahah_helper.module b/ahah_helper.module
index 018da14..75e78c3 100644
--- a/ahah_helper.module
+++ b/ahah_helper.module
@@ -151,8 +151,7 @@ function ahah_helper_render($form_item_to_render = FALSE) {
 
   // Get the form from the cache.
   $form = form_get_cache($form_build_id, $form_state);
-  $args = $form['#parameters'];
-  $form_id = array_shift($args);
+  $form_id = (!empty($form['#parameters'])) ? $form['#parameters'][0] : $_POST['form_id'];
 
   // Are we on the node form?
   $node_form = FALSE;
