Index: ahah_helper.module
===================================================================
RCS file: /cvs/drupal-contrib/contributions/modules/ahah_helper/ahah_helper.module,v
retrieving revision 1.4
diff -u -p -r1.4 ahah_helper.module
--- ahah_helper.module	18 Sep 2009 15:09:07 -0000	1.4
+++ ahah_helper.module	30 Dec 2010 21:32:11 -0000
@@ -197,7 +197,7 @@ function ahah_helper_render($form_item_t
   drupal_json(array(
     'status'   => TRUE,
     'data'     => theme('status_messages') . drupal_render($form_item),
-    'settings' => array('ahah' => $settings['ahah']),
+    'settings' => $settings,
   ));
 }
 
