diff --git a/ahah_helper.module b/ahah_helper.module
index dfd222a..018da14 100644
--- a/ahah_helper.module
+++ b/ahah_helper.module
@@ -236,7 +236,7 @@ function ahah_helper_render($form_item_to_render = FALSE) {
   drupal_json(array(
     'status'   => TRUE,
     'data'     => theme('status_messages') . drupal_render($form_item),
-    'settings' => array('ahah' => $settings['ahah']),
+    'settings' => $settings,
   ));
 }
 
