diff --git a/plugins/context_reaction_block.inc b/plugins/context_reaction_block.inc
index 14c191a..7e6a4ed 100644
--- a/plugins/context_reaction_block.inc
+++ b/plugins/context_reaction_block.inc
@@ -237,9 +237,6 @@ class context_reaction_block extends context_reaction {
       global $user;
       if (module_exists('jquery_ui') && $user->uid) {
         $editable = TRUE;
-        jquery_ui_add(array('ui.draggable', 'ui.droppable', 'ui.sortable'));
-        drupal_add_js(drupal_get_path('module', 'context') .'/plugins/context_reaction_block.js');
-        drupal_add_css(drupal_get_path('module', 'context') .'/plugins/context_reaction_block.css');
       }
       else {
         $editable = FALSE;
