diff --git a/plugins/.context_reaction_block.inc.swp b/plugins/.context_reaction_block.inc.swp
index 4fc689f..92c4148 100644
Binary files a/plugins/.context_reaction_block.inc.swp and b/plugins/.context_reaction_block.inc.swp differ
diff --git a/plugins/context_reaction_block.inc b/plugins/context_reaction_block.inc
index cecdf6c..9ce5ac5 100644
--- a/plugins/context_reaction_block.inc
+++ b/plugins/context_reaction_block.inc
@@ -115,7 +115,7 @@ class context_reaction_block extends context_reaction {
   function editor_form($context) {
     $form = array();
     if (module_exists('jquery_ui')) {
-      jquery_ui_add(array('ui.draggable', 'ui.droppable', 'ui.sortable'));
+      jquery_ui_add(array('ui.draggable', 'ui.droppable', 'ui.sortable'), '1.8');
       drupal_add_js(drupal_get_path('module', 'context_ui') .'/json2.js');
       drupal_add_js(drupal_get_path('module', 'context') .'/plugins/context_reaction_block.js');
       drupal_add_css(drupal_get_path('module', 'context') .'/plugins/context_reaction_block.css');
@@ -237,7 +237,7 @@ class context_reaction_block extends context_reaction {
       global $user;
       if (module_exists('jquery_ui') && $user->uid) {
         $editable = TRUE;
-        jquery_ui_add(array('ui.draggable', 'ui.droppable', 'ui.sortable'));
+        jquery_ui_add(array('ui.draggable', 'ui.droppable', 'ui.sortable'), '1.8');
         drupal_add_js(drupal_get_path('module', 'context') .'/plugins/context_reaction_block.js');
         drupal_add_css(drupal_get_path('module', 'context') .'/plugins/context_reaction_block.css');
       }
